### ISSUE TITLE
Unexpected plan output of aws_s3_bucket module's versioning attribute 

### ISSUE THREAD

**Issue by arunvelsriram** [hashicorp/terraform#16076]
### Terraform Version
Terraform v0.10.4

### Terraform Configuration Files

**modules/s3.tf**

```
variable "bucket_name" { }

variable "versioning" {
  default = []
}

variable "tags" {
  default = {}
}

resource "aws_s3_bucket" "bucket" {
  bucket = "${var.bucket_name}"
  versioning = "${var.versioning}"
  tags = "${var.tags}"
}

```

**test.tf**
```
module "test_bucket" {
  source = "modules"
  bucket_name = "test-bucket-777"
  tags {
    Purpose = "testing"
    Environment = "preprod"
  }
}
```

### Execute
```
terraform plan
terraform apply
terraform plan
```

### Expected Behavior

```
No changes. Infrastructure is up-to-date.
```

### Actual Behavior
```
Terraform will perform the following actions:

  ~ module.test_bucket.aws_s3_bucket.bucket
      versioning.#: "1" => "0"

Plan: 0 to add, 1 to change, 0 to destroy.
```
Regardless of however many times `terraform apply` is executed (and no changes are done after),  `terraform plan` gives the above output.

And the tfstate contains:

```
...
"versioning.#": "1",
"versioning.0.enabled": "false",
"versioning.0.mfa_delete": "false",
...
```

I don't understand why *versioning* needs to be a list (its going to have one element always).

Am able to workaround this by modifying the module as follows:

**modules/s3.tf**

```
variable "bucket_name" { }

variable "versioning" {
  default = {}
}

variable "tags" {
  default = {} # set default value to empty map
}

resource "aws_s3_bucket" "bucket" {
  bucket = "${var.bucket_name}"
  versioning = ["${var.versioning}"] # use list here
  tags = "${var.tags}"
}

```

This is working because now *versioning* list is not empty anymore and contains an empty map by default.

**Comment by jbardin** [comment]
Hi @arunvelsriram,

Yes, the `versioning` field is a list, but because of the way the schema works, it's technically a list of Resource types, which can't technically be translated to the primitive types in variables. 

As you can see a list of maps does seem to work, but that's mostly by chance right now. We have plans to address edge cases like this in future configuration work.

**Comment by apparentlymart** [comment]
Hi @arunvelsriram! Sorry for the long silence here.

In the forthcoming v0.12.0 release we're introducing a new dynamic child block construct that gives an officially-supported way to conditionally declare blocks.

I created the following variant of your example to verify with the Terraform v0.12.0-alpha1 build:

```hcl
provider "aws" {
  region = "us-west-2"
}

variable "bucket_name" {
  type    = string
  default = "tf16076-test"
}

variable "versioning" {
  type = object({
    enabled    = bool
    mfa_delete = bool
  })
  default = {
    enabled    = false
    mfa_delete = false
  }
}

variable "tags" {
  type    = map(string)
  default = {}
}

resource "aws_s3_bucket" "bucket" {
  bucket = var.bucket_name
  tags   = var.tags

  dynamic "versioning" {
    # Generate "versioning" block only if var.versioning.enabled is set.
    # This list will be empty if not, due to the "if" clause.
    for_each = [for x in [var.versioning]: x if x.enabled]
    content {
      enabled    = versioning.value.enabled
      mfa_delete = versioning.value.mfa_delete
    }
  }
}
```

With this configuration as-is, I got the following plan:

```
Terraform will perform the following actions:

  # aws_s3_bucket.bucket will be created
  + resource "aws_s3_bucket" "bucket" {
      + acceleration_status         = (known after apply)
      + acl                         = "private"
      + arn                         = (known after apply)
      + bucket                      = "tf16076-test"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

After changing the `versioning` variable to have `enabled = true` instead, the `versioning` block is dynamically added:

```
Terraform will perform the following actions:

  # aws_s3_bucket.bucket will be created
  + resource "aws_s3_bucket" "bucket" {
      + acceleration_status         = (known after apply)
      + acl                         = "private"
      + arn                         = (known after apply)
      + bucket                      = "tf16076-test"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + versioning {
          + enabled    = true
          + mfa_delete = false
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

This example shows that the readability of this is not ideal for this situation where there's either zero or one blocks of a given type, rather than an arbitrary-sized list of blocks, but the new `for` expressions syntax at least allows that to be expressed reasonably concisely. We may iterate on this further in future releases, but this does at least meet the use-case of allowing a module to conditionally produce a `versioning` block.

It's also possible for the module caller to leave the `mfa_delete` flag as `null` to avoid restating the default of `false`; in v0.12, in resource blocks `null` is always equivalent to not setting it at all.

As James mentioned in his earlier comment, the reason for the weird behavior in v0.11 here was that `versioning` is a nested block rather than an attribute and so in v0.11 and prior there was no officially-supported way to set it conditionally. Treating it like an attribute and assigning a list of maps to it only worked in some edge cases by accident, due to the similarities of how nested blocks were represented internally. This new `dynamic` block is the official way to generate nested blocks dynamically from expressions, and since it is itself built from block syntax rather than attribute syntax it can support all of the same functionality that a static block can support.

Since this use-case is now met in the master branch, I'm going to close this out. This new feature will be included in the forthcoming v0.12.0 final release. Thanks for reporting this!
